WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -1,130 +1,67 @@
 using WalletWasabi.WabiSabi.Models;
 using WalletWasabi.WabiSabi.Models.Decomposition;
 
 namespace WalletWasabi.WabiSabi.Client
 {
 	public class AmountDecomposer
 	{
-		public AmountDecomposer(FeeRate feeRate, Money minAllowedOutputAmount, int outputSize, int availableVsize)
+		public AmountDecomposer(IEnumerable<SmartCoin> possibleInputCoins, RoundState roundState, int maxAvailableVsize)
 		{
-			OutputSize = outputSize;
-			FeeRate = feeRate;
-			AvailableVsize = availableVsize;
-			MinimumAmountPlusFee = minAllowedOutputAmount + OutputFee;
-			StandardDenominationsPlusFee = StandardDenomination.Values
-				.Select(x => x + OutputFee)
-				.OrderByDescending(x => x)
-				.ToImmutableArray();
-		}
-
-		public FeeRate FeeRate { get; }
-		public int AvailableVsize { get; }
-		public Money MinimumAmountPlusFee { get; }
-		public Money OutputFee => FeeRate.GetFee(OutputSize);
-		public int OutputSize { get; }
-		private ImmutableArray<Money> StandardDenominationsPlusFee { get; }
-
-		public IEnumerable<Money> Decompose(IEnumerable<Coin> myInputCoins, IEnumerable<Coin> allInputCoins)
-		{
-			var histogram = GetDenominationProbabilities(allInputCoins);
-
-			var inputs = myInputCoins.Select(x => x.EffectiveValue(FeeRate));
-			var remaining = inputs.Sum();
-
-			var denoms = histogram
-				.OrderByDescending(x => x.Key)
-				.Where(x => x.Value > 1)
-				.Select(x => x.Key)
-				.ToArray();
+			RoundState = roundState;
 
-			List<Money> outputAmounts = new();
-			var remainingVsize = AvailableVsize;
+			// TODO limit to round imposed value
+			var maxEffectiveValue = possibleInputCoins.Sum(x => x.EffectiveValue(RoundState.FeeRate));
 
-			bool end = false;
-			foreach (var denom in denoms.Where(x => x <= remaining))
-			{
-				while (denom <= remaining)
-				{
-					if (remaining < MinimumAmountPlusFee || remainingVsize < 2 * OutputSize)
-					{
-						end = true;
-						break;
-					}
+			var allowedStandardValues = StandardDenomination.Values.Where(x => roundState.CoinjoinState.Parameters.AllowedOutputAmounts.Contains(x));
 
-					outputAmounts.Add(denom - OutputFee);
-					remaining -= denom;
-					remainingVsize -= OutputSize;
-				}
+			// TODO allow more than 6 outputs... but when?
+			var maxOutputs = Math.Min(maxAvailableVsize / Constants.P2WPKHOutputSizeInBytes, 6);
 
-				if (end)
-				{
-					break;
-				}
-			}
+			Logger.LogDebug($"Computing possible decompositions up to {maxOutputs} outputs ({maxAvailableVsize} vbytes available) for {maxEffectiveValue}");
 
-			if (remaining >= MinimumAmountPlusFee)
-			{
-				outputAmounts.Add(remaining - OutputFee);
-			}
+			PossibleDecompositions = new PossibleDecompositions(
+				allowedStandardValues,
+				maxEffectiveValue, // TODO needs to be higher to evaluate other users' inputs
+				Money.Zero,
+				maxOutputs);
 
-			return outputAmounts;
+			Logger.LogDebug($"At {FeeRate.Zero}, best decomposition is {string.Join(' ', PossibleDecompositions.GetByTotalValue().First().Outputs)}");
```

Comment:
```suggestion
			Logger.LogDebug($"At {FeeRate.Zero}, best decomposition is {string.Join(' ', PossibleDecompositions.GetByTotalValue().First().Outputs)}.");
```

Reply:
`FeeRate.Zero`: I don't understand this. Why are we using 0 sat/B here?

Reply:
`FeeRate.Zero`: I don't understand this. Why are we using 0 sat/B here?

Reply:
@nothingmuch tagging because not sure if you will be able to see this comment